x/cronos/keeper/evm.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	evmtypes "github.com/tharsis/ethermint/x/evm/types"
+
+	"github.com/crypto-org-chain/cronos/x/cronos/types"
+)
+
+// CallEVM execute an evm message from native module
+func (k Keeper) CallEVM(ctx sdk.Context, to *common.Address, data []byte, value *big.Int) (*ethtypes.Message, *evmtypes.MsgEthereumTxResponse, error) {
+	k.evmKeeper.WithContext(ctx)
+
+	nonce := k.evmKeeper.GetNonce(types.EVMModuleAddress)
+	msg := ethtypes.NewMessage(
+		types.EVMModuleAddress,
+		to,
+		nonce,
+		value,         // amount
+		25000000,      // gasLimit, FIXME configuration
+		big.NewInt(0), // gasPrice
+		data,
+		nil,   // accessList
+		false, // checkNonce
+	)
+
+	params := k.evmKeeper.GetParams(ctx)
+	// return error if contract creation or call are disabled through governance
+	if !params.EnableCreate && to == nil {
+		return nil, nil, errors.New("failed to create new contract")
+	} else if !params.EnableCall && to != nil {
+		return nil, nil, errors.New("failed to call contract")
+	}
+	ethCfg := params.ChainConfig.EthereumConfig(k.evmKeeper.ChainID())
+
+	// get the coinbase address from the block proposer
+	coinbase, err := k.evmKeeper.GetCoinbaseAddress(ctx)
+	if err != nil {
+		return nil, nil, errors.New("failed to obtain coinbase address")
+	}
+	evm := k.evmKeeper.NewEVM(msg, ethCfg, params, coinbase, nil)
+	ret, err := k.evmKeeper.ApplyMessage(evm, msg, ethCfg, true)
+	if err != nil {
+		return nil, nil, err
+	}
+	k.evmKeeper.CommitCachedContexts()
+	return &msg, ret, nil
+}
+
+// CallCronosERC20 call a method of CronosERC20 contract
+func (k Keeper) CallCronosERC20(ctx sdk.Context, contract common.Address, method string, args ...interface{}) ([]byte, error) {
+	data, err := types.CronosERC20Contract.ABI.Pack(method, args...)
+	if err != nil {
+		return nil, err
+	}
+	_, res, err := k.CallEVM(ctx, &contract, data, big.NewInt(0))
+	if err != nil {
+		return nil, err
+	}
+	if res.Failed() {
+		return nil, fmt.Errorf("call contract failed: %s, %s, %s", contract.Hex(), method, res.Ret)
+	}
+	return res.Ret, nil
+}
+
+// DeployCronosERC20 deploy an embed erc20 contract
+func (k Keeper) DeployCronosERC20(ctx sdk.Context, denom string) (common.Address, error) {
+	ctor, err := types.CronosERC20Contract.ABI.Pack("", denom, uint8(0))
+	if err != nil {
+		return common.Address{}, err
+	}
+	data := append(types.CronosERC20Contract.Bin, ctor...)
+
+	msg, res, err := k.CallEVM(ctx, nil, data, big.NewInt(0))
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	if res.Failed() {
+		return common.Address{}, fmt.Errorf("contract deploy failed: %s", res.Ret)
+	}
+	return crypto.CreateAddress(types.EVMModuleAddress, msg.Nonce()), nil
+}
+
+// SendCoinFromNativeToERC20 convert native token to erc20 token
+func (k Keeper) SendCoinFromNativeToERC20(ctx sdk.Context, sender common.Address, coin sdk.Coin) error {
```

Comment:
`ConvertToEVMTokens` or `ConvertToERC20` to make it more generic?

Reply:
The name `SendCoinFromXToY` is following [bank keeper's style](https://github.com/cosmos/cosmos-sdk/blob/master/x/bank/keeper/keeper.go#L34), the ERC20 represent the ABI, we can replace it with the token standard name we chose in the end.

Reply:
Yeah but this function is not only for `Sending` Coin right?

It does multiple things (sending coin from user account to contract account, deploy contract, mint tokens). I think using the term `Convert` or `Wrapping` would be more accurate

Reply:
Sounds good, changed to `ConvertNativeToERC20/ConvertERC20ToNative`.

Reply:
Oh, we still have the `SendCoin` and `SendCoins` variations, which are not covered by the `Convert*` name, so maybe `ConvertCoinFromNativeToERC20`/`ConvertCoinsFromNativeToERC20`, which is very close to the original one 😄